examples/hackweek_demo/plot_pnl.py

```diff
@@ -112,93 +113,134 @@ def get_wallet_from_onchain_trade_info(
+            state = pool_info.loc[block]
+            if block in checkpoint_info.index:  # a checkpoint exists
+                current_checkpoint = checkpoint_info.loc[block]
+                maturity = current_checkpoint["timestamp"] + pd.Timedelta(seconds=position_duration)
+            else:
+                maturity = None
+
             spot_price = calculate_spot_price(
-                row.share_reserves,
-                row.bond_reserves,
-                row.lp_total_supply,
-                row.maturity_time,
-                row.block_timestamp,
+                state.shareReserves,
+                state.bondReserves,
+                state.lpTotalSupply,
+                maturity,
+                current_wallet.timestamp,
                 position_duration,
             )
-            # print(f"{spot_price=}")
 
-            # LP value (TODO: check why certain agents own more than 100% of the pool)
-            total_lp_value = row.share_reserves / 1e18 * row.share_price / 1e18 + row.bond_reserves / 1e18 * spot_price
-            share_of_pool = float(agent_wallets[agent].lp_tokens) / (row.lp_total_supply / 1e18)
-            # print(f"{float(agent_wallets[agent].lp_tokens)=}")
-            # print(f"{row.lp_total_supply/1e18=}")
-            pnl = share_of_pool * total_lp_value
-
-            # for each LONG
-            for _, long in agent_wallets[agent].longs.items():
-                pnl += float(long.balance) * spot_price
-
-            # for each SHORT
-            for _, short in agent_wallets[agent].shorts.items():
-                pnl += float(short.balance) * (1 - spot_price)
-
-            new_pnl[f"{agent_shortname}"] = pnl
-        pnl_data.append(new_pnl)
-        time_data.append(row["timestamp"])
+            pnl = 0
+            for position_name in positions.columns:
+                if position_name.startswith("LP"):
+                    # LP value (TODO: check why certain agents own more than 100% of the pool)
+                    total_lp_value = state.shareReserves * state.sharePrice + state.bondReserves * spot_price
+                    share_of_pool = position.LP / state.lpTotalSupply
+                    assert share_of_pool < 1, "share_of_pool must be less than 1"
+                    pnl += share_of_pool * total_lp_value
+                elif position_name.startswith("LONG"):
+                    # LONG value
+                    pnl += position.loc[position_name] * spot_price
+                elif position_name.startswith("SHORT"):
+                    # SHORT value is calculated as the:
+                    # total amount paid for the position (position * 1)
+                    # minus the closing cost (position * spot_price)
+                    # this equals position * (1 - spot_price)
+                    pnl += position.loc[position_name] * (1 - spot_price)
+
+            wallet.loc[block, "pnl"] = pnl
+        agent_info = wallet, positions, deltas, open_share_price
```

Comment:
why are we overwriting agent_info here?

Reply:
replacing the wallet info in the dict[agent, wallet_info] with dict[agent, lots_more_stuff_than_that]